src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -180,11 +190,44 @@ public Task StopAsync()
             catch (Exception ex)
             {
                 Log.UnexpectedExceptionInStop(Logger, ConnectionId, ex);
-            }
-            
+            }
+
             return Task.CompletedTask;
         }
 
+        public async Task CloseAsync()
+        {
+            if (Interlocked.CompareExchange(ref _closeStage, ClosingSendingFin1, ClosingDefault) != ClosingDefault)
+            {
+                // CloseAsync is in progress or done.
+                return;
+            }
+
+            while (Interlocked.CompareExchange(ref _closeStage, ClosingWaitingApplicationTasks, ClosingReceivedFinAck1) != ClosingReceivedFinAck1)
+            {
+                await Task.Delay(TimeSpan.FromMilliseconds(50));
+            }
+
+            if (clientOutgoingTasks.Values.Count > 0)
+            {
+                await Task.WhenAll(clientOutgoingTasks.Values);
+            }
+
+            if (_connectionContext != null)
```

Comment:
What is this for?

Reply:
We're not sure if the buffer is empty even after all outgoing tasks have done.

So we call the `Flush` method explicitly to guarantee all messages can be sent to `ASRS`.

Reply:
This doesn't look right. We should have a single place that flushes and this should be awaiting that place.

Reply:
You're right. 
The `incomingPipe` is no longer valid after all outgoing tasks are done. 
So I can just Complete the `incomingPipe` and wait the connection stopped automatically.